tiledb/sm/cpp_api/error.h

```diff
@@ -90,7 +90,11 @@ class Error {
   const std::string error_message() {
     const char* msg = nullptr;
     tiledb_error_message(error_.get(), &msg);
-    return std::string(msg);
+    if (msg == nullptr)
+      return std::string(
```

Comment:
I believe we should return empty string here. This matches the c-api closer.

Reply:
I have no dog in this fight, and suggested this too. Isaiah's reasoning was that "whenever this is called we know there is an error" and a properly formatted text is better.  Either one works for me and is better that what I get without it (at the R side when no error has occurred).

Reply:
I'm fine with changing it if you prefer @Shelnutt2, but my reasoning is that we should never be in an error state without an error message, so reaching this state is extraordinary. I would rather get _some message_ that indicates a small amount of debug information rather than an empty message.

If my assumption that we should never (in principle) have an error state without setting an error message is incorrect, then disregard.

Reply:
I agree with @ihnorton here, so that we can find what caused this error more easily in the code (especially if there are multiple such code paths that used to return empty strings).

Reply:
From my testing via the R API built on top, and where I just added an error message accessor, it seems like this:

- when we not yet had any error, this situation arises; it currently does not create a 'nice' (or even empty string)
- once we had an error, the most recent error message is appropriately cached and available
- even when the last operation was successful, the error message getter the most recent eror (which may not correspond to the most recent operation, but that is a different issue if at all)

Reply:
This function can be called arbitrarily, not just from an error state. In the C++ API error states throw exceptions. The motivation for adding this API was the `Context::last_error()` API so users (we saw a use case in TileDB-R) could get an error from the context higher up in the application stack than where the exception might be thrown and caught.

If we do want to keep a friendly "no error" message we need to change the c-api also. We should not have the c-api return nullptr but c++ return a message.

Reply:
Ok, don't think it's worth changing the C API. `return std::string("")` SGTM.

Reply:
Coming full circle as that is where we started.